[GlobalISel] Make scalar G_SHUFFLE_VECTOR illegal.

davemgreen opened the pull request:
I'm not sure if this is the best way forward or not, but we have a lot of issues with forgetting that shuffle_vectors can be scalar again and again. (There is another example from the recent known-bits code added recently). As a scalar-dst shuffle vector is just an extract, and a scalar-source shuffle vector is just a build vector, this patch makes scalar shuffle vector illegal and adjusts the irbuilder to create the correct node as required. 

Most targets do this already through lowering or combines. Making scalar shuffles illegal simplifies gisel as a whole, it just requires that transforms that create shuffles of new sizes to account for the scalar shuffle being illegal (mostly IRBuilder and LessElements).

aemerson commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -3330,6 +3330,51 @@ bool IRTranslator::translateShuffleVector(const User &U,
     Mask = SVI->getShuffleMask();
   else
     Mask = cast<ConstantExpr>(U).getShuffleMask();
+
```
For future readers some comments here would be useful to quickly parse that the code is trying to do.

aemerson reviewed (commented):
I don't have a problem with this in principle, but let's check with @qcolombet in case there's something we're missing.

arsenm commented:
This also came up recently in https://github.com/llvm/llvm-project/pull/139769

arsenm commented on llvm/test/CodeGen/MIR/AArch64/parse-shufflemask.mir:
```diff
@@ -122,54 +122,3 @@ body:             |
-
-    %0:_(<2 x s32>) = COPY $d0
-    %2:_(<2 x s32>) = G_IMPLICIT_DEF
-    %1:_(s32) = G_SHUFFLE_VECTOR %0(<2 x s32>), %2, shufflemask(0)
```
These cases should still be tested for emitting an error

lialan commented on llvm/lib/CodeGen/GlobalISel/MachineIRBuilder.cpp:
```diff
@@ -784,9 +784,9 @@ MachineInstrBuilder MachineIRBuilder::buildShuffleVector(const DstOp &Res,
-  const LLT DstElemTy = DstTy.isVector() ? DstTy.getElementType() : DstTy;
-  const LLT ElemTy1 = Src1Ty.isVector() ? Src1Ty.getElementType() : Src1Ty;
-  const LLT ElemTy2 = Src2Ty.isVector() ? Src2Ty.getElementType() : Src2Ty;
+  const LLT DstElemTy = DstTy.getScalarType();
```
When building a new `ShuffleVector`, should check that `len(mask) > 1` to avoid accidentally yielding a new scalar.

qcolombet commented:
> I don't have a problem with this in principle, but let's check with @qcolombet in case there's something we're missing.

Sounds good to me.

Pierre-vh reviewed (approved):
LGTM if nits addressed, after fixing them wait a bit to see if other reviewers have objections too

davemgreen committed:
[GlobalISel] Make scalar G_SHUFFLE_VECTOR illegal.

I'm not sure if this is the best design or not, but we have a lot of issues
with forgetting that shuffle_vectors can be scalar again and again. (There is
another example from the recent known-bits code added recently). As a
scalar-dst shuffle vector is just an extract, and a scalar-source shuffle
vector is just a build vector, this patch makes scalar shuffle vector illegal
and adjusts the irbuilder to create the correct node as required.
